### PR TITLE
New resource meta-attribute "allow-unhealthy-nodes"

### PR DESCRIPTION
**Pull request by kgaillot**

### CODE REVIEW

**Comment by kgaillot** [comment]
@clumens , can you review this when you get a chance? Thanks

The actual feature is tiny, this is mostly clean-up

File: lib/pengine/common.c [hunk 1]
```diff
@@ -240,40 +245,45 @@ static pcmk__cluster_option_t pe_opts[] = {
...
+        PCMK__VALUE_NONE ", " PCMK__VALUE_MIGRATE_ON_RED ", "
+            PCMK__VALUE_ONLY_GREEN ", " PCMK__VALUE_PROGRESSIVE ", "
+            PCMK__VALUE_CUSTOM,
+        PCMK__VALUE_NONE, check_health,
         "How cluster should react to node health attributes",
         "Requires external entities to create node attributes (named with "
-            "the prefix \"#health\") with values \"red\", \"yellow\" or "
-            "\"green\"."
+            "the prefix \"#health\") with values \"" PCMK__VALUE_RED "\", "
+            "\"" PCMK__VALUE_YELLOW "\", or " "\"" PCMK__VALUE_GREEN "\"."
```
Comment (clumens) [inline lib/pengine/common.c:255]: I think you've got a few too many quotes after the "or".  I think that bit can just be `, or \""`

File: lib/common/health.c [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+                            pcmk__str_casei)) {
+        return pcmk__health_strategy_custom;
+
+    } else {
+        pcmk__config_err("Using default of \"" PCMK__VALUE_NONE "\" for "
+                         PCMK__OPT_NODE_HEALTH_STRATEGY
+                         " because '%s' is not a valid value",
+                         value);
+        return pcmk__health_strategy_none;
+    }
```
Comment (clumens) [inline lib/common/health.c:49]: It'd be nice to have unit tests for easy new stuff like this.

File: lib/pengine/common.c (hunk 1 above)
Comment (kgaillot) [reply lib/pengine/common.c:255]: good catch, updated commit

File: lib/common/health.c (hunk 2 above)
Comment (kgaillot) [reply lib/common/health.c:49]: good point, added "unit test" commit

**Comment by wenningerk** [comment]
Maybe slightly related as it might result in some kind of a generalization of the approach to allow certain resources to be started when/where we usually wouldn't:
With certain fencing-approaches it from time to time appears that they need some resources to be operational to be able to give a proper result (communication layers and alike).
So some kind of flag that makes pacemaker start a resource prior to cluster forming (including fencing of everybody unseen) might make sense - if handled with care.

**Comment by clumens** [comment]
Looks good to me with the changes.

**Comment by kgaillot** [comment]
> It occurs to me we could overload requires for the health exemption but it would be a little tricky since it would now be a combination of things like requires=fencing,health. Thoughts?

I like the logical association here, but I wouldn't want to have to write the documentation for the default and allowed values ;)

The current documentation for the requires default is already ridiculous: "quorum for resources with a class of stonith, otherwise unfencing if unfencing is active in the cluster, otherwise fencing if stonith-enabled is true, otherwise quorum"

**Comment by kgaillot** [comment]
Updated last commit to add "since 2.1.3" to documentation

**Comment by wenningerk** [comment]
> quorum for resources with a class of stonit

Speaking of documentation ... that part is obviously hit especially badly by pdf-generation-issues in
https://clusterlabs.org/pacemaker/doc/2.1/Pacemaker_Explained/pdf/Pacemaker_Explained.pdf

Room in the table might be a little squeezed but otherwise a graphical approach - flow-chart-like - might be a way to document it in a way that is easier to consume.
Basically I like the idea of using requires for health as well.
When depending on such requires=nothing resources with a fencing-device we are probably back to where fenced isn't updating that info properly once the resource is running - right?
But might still be useful to just try as with this kind of fencing-devices we should anyway have some topology to fall through to some other fencing-means. Would just be cleaner not to try if we theoretically could know it is doomed beforehand.

**Comment by kgaillot** [comment]
Ugh. I like the logic of "requires" as a list, and it would work fine when left to default, but it wouldn't work with existing explicit configurations. E.g. if someone has already set requires="fencing" they probably don't intend to exempt the resource from health checks.

We could flip the sense so the user has to explicitly list a health exemption, but that wouldn't read logically. E.g. requires="fencing,unhealthy-ok"

It could possibly be done with a CIB schema major version bump using an XSL transform to add "health" to any existing explicit "requires", but I wouldn't want to do that a minor-minor release, and it's rather soon for another major or minor release.

I think we have to chalk that up to something that would have been nice in the original design, but stick with allow-unhealthy-nodes at this point.

**Comment by kgaillot** [comment]
> > quorum for resources with a class of stonit
> 
> Speaking of documentation ... that part is obviously hit especially badly by pdf-generation-issues in https://clusterlabs.org/pacemaker/doc/2.1/Pacemaker_Explained/pdf/Pacemaker_Explained.pdf

Yeah, sphinx's PDF generation isn't very good. Dr. Google suggests adding ":class: longtable" to a table that needs to span multiple pages, so I'll try that.

> 
> Room in the table might be a little squeezed but otherwise a graphical approach - flow-chart-like - might be a way to document it in a way that is easier to consume.

For some other complicated defaults we've put the description in a separate section, and the table just says "See <section>". That might work with that approach.

**Comment by wenningerk** [comment]
> Ugh. I like the logic of "requires" as a list, and it would work fine when left to default, but it wouldn't work with existing explicit configurations. E.g. if someone has already set requires="fencing" they probably don't intend to exempt the resource from health checks.
> 
> We could flip the sense so the user has to explicitly list a health exemption, but that wouldn't read logically. E.g. requires="fencing,unhealthy-ok"

Alternative might be some way to explicitly state that what you are setting is meant to be a list.
Unfortunately I don't have an idea of a convention that could be adopted.
One - though not 100% intuitive - possibility would be having a colon in it to show it is a list - if just a single element then with a trailing colon.

**Comment by kgaillot** [comment]
> One - though not 100% intuitive - possibility would be having a colon in it to show it is a list - if just a single element then with a trailing colon.

That's tempting, but then it would be different from every other list option, and I think that would bite us later.

A similar approach might be a "not" operator, like: requires="fencing,!health" but I think a separate meta-attribute will be easier for users to understand.

**Pull request merged**